crates/cli/src/suite.rs

```diff
@@ -0,0 +1,139 @@
+    fn as_ref(&self) -> &OsStr {
+        match self {
+            BenchmarkOrSuite::Suite(suite) => suite.path.as_os_str(),
+            BenchmarkOrSuite::Benchmark(path) => path.as_os_str(),
+        }
+    }
+}
+
+/// Parse a [BenchmarkOrSuite] from a string path; files ending in `.suite` are
+/// assumed to be suite files.
+impl FromStr for BenchmarkOrSuite {
+    type Err = anyhow::Error;
+    fn from_str(path: &str) -> Result<Self, Self::Err> {
+        Ok(if Suite::has_extension(path) {
+            Self::Suite(Suite::parse(path)?)
+        } else {
+            Self::Benchmark(PathBuf::from(path))
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct Suite {
+    path: PathBuf,
+    benchmarks: Vec<PathBuf>,
+}
+
+impl Suite {
+    /// Parse the contents of a suite file:
+    /// - empty lines are ignored
+    /// - `#`-prefixed lines are ignored
+    /// - relative paths are resolved using the parent directory of the
+    ///   `suite_path`.
+    pub fn parse<P: AsRef<Path>>(suite_path: P) -> Result<Self> {
+        Self::parse_contents(suite_path.as_ref(), &fs::read(suite_path.as_ref())?)
+    }
+
+    /// Utility function for easier testing.
+    fn parse_contents<P: AsRef<Path>>(suite_path: P, file_contents: &[u8]) -> Result<Self> {
+        use io::BufRead;
+        let suite_path = suite_path.as_ref().to_path_buf();
+        let parent_dir = suite_path
+            .parent()
+            .expect("the suite path must have a parent directory");
+        let mut benchmarks = vec![];
+        for line in io::BufReader::new(file_contents)
+            .lines()
+            .filter_map(Result::ok)
+        {
+            let line = line.trim();
+            if !line.starts_with("#") && !line.is_empty() {
+                benchmarks.push(parent_dir.join(line))
+            }
+        }
+        Ok(Self {
+            path: suite_path,
+            benchmarks,
+        })
+    }
+
+    /// Check if a path has the `suite` extension.
+    fn has_extension<P: AsRef<Path>>(path: P) -> bool {
+        match path.as_ref().extension() {
+            Some(ext) => ext == "suite",
+            None => false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const SUITE_PATH: &str = "/home/bench.suite";
+    const CONTENTS: &str = "
+        # These benchmarks are...
+        a.wasm
+        /b.wasm
+        ../c.wasm
+    ";
+
+    #[test]
+    fn parse() {
+        let suite = Suite::parse_contents(SUITE_PATH, CONTENTS.as_bytes()).unwrap();
+
+        // The suite path should be preserved.
+        assert_eq!(suite.path, PathBuf::from(SUITE_PATH));
+
+        assert_eq!(
+            suite.benchmarks,
+            vec![
+                // Relative paths are appended.
+                PathBuf::from("/home/a.wasm"),
+                // Absolute paths are preserved.
+                PathBuf::from("/b.wasm"),
+                // Canonicalization happens later.
+                PathBuf::from("/home/../c.wasm")
```

Comment:
FWIW, we shouldn't ever call `canonicalize` on the path because it just does super weird stuff on windows that isn't really what we want, and using non-canonical paths doesn't hurt for our use cases in sightglass.

Reply:
I made a note of this in the documentation just to remember why to keep these paths non-canonicalized. I chose to keep them non-canonicalized out of testing laziness; this is a more valid concern.